### PR TITLE
[pkg/loki] Group Loki requests by the attribute inferred from the `loki.tenant` hint.

### LINKED ISSUE #14706: Possibility to set Loki tenant id from resource attributes
### Describe the issue you're reporting

Currently the Loki exporter supports setting a dynamic tenant id based on a resource attribute using the tenant field. This field is deprecated and will be removed in v0.61.

The suggested alternative to use the "headers_setter" does not support resource attributes.

As proposed by @jpkrohling on Slack, something like a "loki.attributes.tenant" label could be introduced using a processor which can be picked up by the exporter to be set as tenant in the header.

### PR DESCRIPTION
**Pull request by kovrus**
**Description:** <Describe what has changed.>

Group Loki requests by an attribute inferred from the `loki.tenant` hint. 

Follow up: use `LogsToLokiRequests` in `exporter/loki`

**Link to tracking Issue:** <Issue number if applicable>
Closes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/14706

**Testing:** <Describe what testing was performed and which tests were added.>
Unit tests

**Documentation:** <Describe the documentation added.>
It will be added in `exporter/loki`.

### CODE REVIEW

File: pkg/translator/loki/logs_to_loki.go [hunk 1]
```diff
@@ -22,13 +22,141 @@ import (
...
+				log := plog.NewLogRecord()
+				logs.At(k).CopyTo(log)
+
+				// we may remove attributes, so we make a copy and change our version
+				resource := pcommon.NewResource()
+				rls.At(i).Resource().CopyTo(resource)
+
+				// resolve tenant and get/create a push request group
+				tenant := getTenantFromTenantHint(log.Attributes(), resource.Attributes())
+				group, ok := groups[tenant]
```
Comment (kovrus) [inline pkg/translator/loki/logs_to_loki.go:74]: The logic is the same as in `LogsToLoki`., only the grouping was added.

File: pkg/translator/loki/logs_to_loki.go [hunk 2]
```diff
@@ -22,13 +22,141 @@ import (
...
+	}
+
+	requests := make(map[string]PushRequest)
+	for tenant, g := range groups {
+		pr := &logproto.PushRequest{
+			Streams: make([]logproto.Stream, len(g.streams)),
+		}
+
+		i := 0
+		for _, stream := range g.streams {
```
Comment (mar4uk) [inline pkg/translator/loki/logs_to_loki.go:121]: why index from the range is not used here? `for i, stream := range g.streams {`

Comment (kovrus) [reply pkg/translator/loki/logs_to_loki.go:121]: `streams` is a map, so `i` is a key that we ignore

**Review by mar4uk — APPROVED** [review]

File: pkg/translator/loki/logs_to_loki.go [hunk 3]
```diff
@@ -22,13 +22,141 @@ import (
...
+		}
+		requests[tenant] = PushRequest{
+			PushRequest: pr,
+			Report:      g.report,
+		}
+	}
+	return requests
+}
+
+// resolveAttributeFromTenantHint extract an attribute based on the tenant hint.
```
Comment (jpkrohling) [inline pkg/translator/loki/logs_to_loki.go:133]:
```suggestion
// getTenantFromTenantHint extract an attribute based on the tenant hint.
```

File: pkg/translator/loki/logs_to_loki_test.go [hunk 4]
```diff
@@ -18,11 +18,295 @@ import (
...
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "use tenant resource attributes if both logs and resource attributes provided",
```
Comment (jpkrohling) [inline pkg/translator/loki/logs_to_loki_test.go:136]: Is this documented? If not, could you please mention it somewhere?

**Review by jpkrohling — APPROVED** [review]

File: pkg/translator/loki/logs_to_loki_test.go (hunk 4 above)
Comment (kovrus) [reply pkg/translator/loki/logs_to_loki_test.go:136]: add a sentence in the method doc.

**Pull request merged**